app/presenters/tree_builder_services.rb

```diff
@@ -36,8 +36,44 @@ def x_get_tree_roots(count_only, _options)
   end
 
   def x_get_tree_custom_kids(object, count_only, _options)
-    # Get My Filters and Global Filters
-    count_only_or_objects(count_only, x_get_search_results(object)) if %w[my global].include?(object[:id])
+    case object[:id]
+    when 'my', 'global'
+      # Get My Filters and Global Filters
+      count_only_or_objects(count_only, x_get_search_results(object))
+    when 'asrv', 'rsrv'
+      retired = object[:id] != 'asrv'
+      services = Rbac.filtered(Service.where(:retired => retired, :display => true))
+      return sevices.size if count_only
+
+      MiqPreloader.preload(services.to_a, :picture)
+      # Cache the tree data to speed up child (count) retrieval
+      @tree_data = Service.arrange_nodes(services.sort_by { |n| [n.ancestry.to_s, n.name.downcase] })
+      @tree_data.keys
+    end
+  end
+
+  def x_get_tree_nested_services(object, count_only)
+    subtree = if @tree_data # If the cached tree data is available, find the child nodes in it
+                deep_find(@tree_data, object).keys
+              else # If it's a lazy load call, we have to retrieve the data manually
+                services = Rbac.filtered(object.descendants.where(:retired => object.retired, :display => true))
+                Service.arrange_nodes(services.sort_by { |n| [n.ancestry.to_s, n.name.downcase] }).keys
```

Comment:
```ruby
# seems like this could be dedup'd maybe?

fetch_services(Service)
fetch_services(object.descendants)

def fetch(query)
  services = query.where(:retired => object.retired, :display => true).to_a
  count = services.size;
  MiqPreloader.preload(services, :picture)
  arranged_services = Service.arrange_nodes(services.sort_by { |n| [n.ancestry.to_s, n.name.downcase] }).keys
  # not totally sure the return value here - but here are some ideas:
  return [services, count, arranged_services]
end
```